main.py

```diff
@@ -486,96 +474,79 @@ def categorise_elongation(elongation: float) -> ElongationCategory:
-#
-#    logged_mape_ys[i] = calculate_mape(properties_array[:,0], predicted_array_1[:,0])
-#    logged_mape_ys[i+1] = calculate_mape(properties_array[:,0], predicted_array_2[:, 0])
-#    logged_mape_ts[i] = calculate_mape(properties_array[:,1], predicted_array_1[:,1])
-#    logged_mape_ts[i+1] = calculate_mape(properties_array[:,1], predicted_array_2[:, 1])
-#    logged_mape_e[i] = calculate_mape(properties_array[:,2], predicted_array_1[:,2])
-#    logged_mape_e[i+1] = calculate_mape(properties_array[:,2], predicted_array_2[:, 2])
-#
-#errors_df = pd.DataFrame([])
-#errors_df["yield_stress_R^2"] = logged_r_squared_ys
-#errors_df["tensile_stress_R^2"] = logged_r_squared_ts
-#errors_df["elongation_R^2"] = logged_r_squared_e
-#errors_df["yield_stress_std"] = logged_std_ys
-#errors_df["tensile_stress_std"] = logged_std_ts
-#errors_df["elongation_std"] = logged_std_e
-#errors_df["yield_stress_mape"] = logged_mape_ys
-#errors_df["tensile_stress_mape"] = logged_mape_ts
-#errors_df["elongation_mape"] = logged_mape_e
-#
-#figure, axis = plt.subplots(3, 3)
-#
-#sns.kdeplot(data=errors_df, x="yield_stress_R^2", clip=(0.0,1.0), ax=axis[0, 0])
-#sns.kdeplot(data=errors_df, x="tensile_stress_R^2", clip=(0.0,1.0), ax=axis[0, 1])
-#sns.kdeplot(data=errors_df, x="elongation_R^2", clip=(0.0,1.0), ax=axis[0, 2])
-#sns.kdeplot(data=errors_df, x="yield_stress_std", clip=(0.0, 500.0), ax=axis[1, 0])
-#sns.kdeplot(data=errors_df, x="tensile_stress_std", clip=(0.0, 500.0), ax=axis[1, 1])
-#sns.kdeplot(data=errors_df, x="elongation_std", clip=(0.0, 12.5), ax=axis[1, 2])
-#sns.kdeplot(data=errors_df, x="yield_stress_mape", clip=(0.0,100.0), ax=axis[2, 0])
-#sns.kdeplot(data=errors_df, x="tensile_stress_mape", clip=(0.0,100.0), ax=axis[2, 1])
-#sns.kdeplot(data=errors_df, x="elongation_mape", clip=(0.0,100.0), ax=axis[2, 2])
-#
-#plt.show()
-#
-#figure, axis = plt.subplots(7, 2)
-#
-#
\ No newline at end of file
+def cross_validation(n, k, data: pd.DataFrame):
+    logged_std_ys = np.zeros(n)
+    logged_std_ts = np.zeros(n)
+    logged_std_e = np.zeros(n)
+    logged_r_squared_ys = np.zeros(n)
+    logged_r_squared_ts = np.zeros(n)
+    logged_r_squared_e = np.zeros(n)
+    logged_mape_ys = np.zeros(n)
+    logged_mape_ts = np.zeros(n)
+    logged_mape_e = np.zeros(n)
+
+    data = data.dropna()
+
+    for i in range(0, n, k - 1):
+        samples: list[pd.DataFrame] = DataframeWriter.data_shuffler(data, k)
+        control_sample: pd.DataFrame = samples[0]
+        properties_array = np.array(control_sample["combined_properties"].values.tolist())
+
+        for j in range(0, k - 1):
+            if i + j >= n:
+                continue
+
+            control_sample[f"combined_predicted_{j + 1}"] = (
+                control_sample["combined_compositions"].apply(lambda prop: np.matmul(a_calc(samples[j + 1]), prop))
+            )
+
+            predicted_array = np.array(control_sample[f"combined_predicted_{j + 1}"].values.tolist())
```

Comment:
I think you might want to remove this creation of columns in the data frame - just extract the values you want directly
```suggestion
            predicted_array = np.array(
                control_sample["combined_compositions"]
                .apply(lambda prop: np.matmul(a_calc(samples[j + 1]), prop))
                .values()
                .tolist()
            )
```

Reply:
Done